### ISSUE TITLE
Problems with scoped lifetime services (AddScoped)

### ISSUE THREAD

**Issue by heikkilamarko** [Azure/azure-functions-host#5098]
Hi,

We are using durable functions with dependency injection. In our `Startup.cs` file, we register our dependencies as follows:

```
services.AddScoped<IMyService>(serviceProvider => new MyService(...))
```

In our activity function, we are using normal constructor injection to get `IMyService` instances.

The problem is that even if we are using `AddScoped` for registering the service, during the activity function run, each class that asks for the service, gets a different instance of `IMyService`. This breaks our app logic, because `IMyService` users won't see each other's changes.

As a workaround to the earlier Azure Functions runtime DI issues, we had the following pinning in place `FUNCTIONS_EXTENSION_VERSION = 2.0.12673.0`. The pinned version is not supported anymore by Azure, so our function DI is now broken again.

**Comment by espray** [comment]
I am afraid this might be by design. You should post this on the Durable Functions board?

**Comment by heikkilamarko** [comment]
I find it hard to believe this is by design. If it is so, documentation should state that very clearly.

**Comment by heikkilamarko** [comment]
The problem is not durable function specific. Same happens with normal `TimerTrigger` functions also.

**Comment by heikkilamarko** [comment]
I made more experiments and it seem that the DI problem occurs only when I have used `services.AddHttpClient()` to register `HttpClients`. If I comment out `services.AddHttpClient()` calls, `AddScoped` works as expected.

**Comment by dhanapalschandran** [comment]
Hi @heikkilamarko I am also facing the same issue, when this issue happens last week I have pinned to older runtime version, then it started working, now again breaking. I agree it's not an issue with durable function. Do you have any workaround for httpclient register or waiting for team to provide global fix

**Comment by unt1tled** [comment]
@dhanapalschandran  Could it be breaking again due to #5096?

**Comment by fabiocav** [comment]
@heikkilamarko can you please share repro code? Is this happening with factory registrations only? Some more details/repro will ensure we're looking exactly at the issue you're running into.

@unt1tled / @dhanapalschandran if you can share the code you're having issues with, we can validate with your scenarios as well.  

Thanks!

**Comment by heikkilamarko** [comment]
Hi. See below for a repro code (see the comment block in the function code):
```
using Microsoft.Azure.Functions.Extensions.DependencyInjection;
using Microsoft.Azure.WebJobs;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Logging;
using System.Net.Http;

[assembly: FunctionsStartup(typeof(Demo.App.Startup))]

namespace Demo.App
{
    public class Startup : FunctionsStartup
    {
        public override void Configure(IFunctionsHostBuilder builder)
        {
            var services = builder.Services;

            services.AddHttpClient<MyHttpClient>();

            services
                .AddScoped<IMyService, MyService>()
                .AddScoped<IMyServiceSetter, MyServiceSetter>()
                .AddScoped<IMyServiceGetter, MyServiceGetter>();
        }
    }

    public class MyHttpClient
    {
        private readonly HttpClient httpClient;

        public MyHttpClient(HttpClient httpClient)
        {
            this.httpClient = httpClient;
        }
    }

    public interface IMyService
    {
        void Set(int value);

        int Get();
    }

    public class MyService : IMyService
    {
        private int value = 0;

        public int Get()
        {
            return value;
        }

        public void Set(int value)
        {
            this.value = value;
        }
    }

    public interface IMyServiceSetter
    {
        void SetValue(int value);
    }

    public class MyServiceSetter : IMyServiceSetter
    {
        private readonly IMyService myService;
        private readonly IMyServiceGetter myServiceGetter;

        public MyServiceSetter(IMyService myService, IMyServiceGetter myServiceGetter)
        {
            this.myService = myService;
            this.myServiceGetter = myServiceGetter;
        }

        public void SetValue(int value)
        {
            myService.Set(value);
        }
    }

    public interface IMyServiceGetter
    {
        int GetValue();
    }

    public class MyServiceGetter : IMyServiceGetter
    {
        private readonly IMyService myService;
        private readonly MyHttpClient myHttpClient;

        public MyServiceGetter(IMyService myService, MyHttpClient myHttpClient)
        {
            this.myService = myService;
            this.myHttpClient = myHttpClient;
        }

        public int GetValue()
        {
            return myService.Get();
        }
    }

    public class DemoFunction
    {
        private readonly IMyServiceSetter myServiceSetter;
        private readonly IMyServiceGetter myServiceGetter;

        public DemoFunction(
            IMyServiceSetter myServiceSetter,
            IMyServiceGetter myServiceGetter)
        {
            this.myServiceSetter = myServiceSetter;
            this.myServiceGetter = myServiceGetter;
        }

        [FunctionName("Demo")]
        public void Demo([TimerTrigger("0 */5 * * * *", RunOnStartup = true)]TimerInfo timerInfo, ILogger log)
        {
            int value1 = 123;
            myServiceSetter.SetValue(value1);
            int value2 = myServiceGetter.GetValue();

            // Logs '123 - 0'
            // It means that MyServiceGetter and MyServiceSetter got different IMyservice instances.
            // Should be the same since we are using AddScoped to register services.
            // ---
            // If you comment out MyHttpClient injection from MyServiceGetter, it logs '123 - 123'.
            log.LogInformation($"{value1} - {value2}");
        }
    }
}

```

**Comment by t-l-k** [comment]
I believe this is the same as my issue #4914 - clearly something wrong with scoped registrations.

@fabiocav can we get an update on these tickets? Are they being investigated? Thx

**Comment by APIWT** [comment]
@fabiocav Hope you are having a nice day :) We are also hopeful that this issue will be fixed soon. We have spent a lot of time writing code to wire up our services and we really love the way scoped services work in ASP.NET Core. Let me know if there is anything I can do to help get this moving!

**Comment by buyckkdelaware** [comment]
We have experienced the same problem, but I also noticed this:
It's only the first time after the function started/triggered, that the scoped services are 'wrong'. The second time, it seems to be fine. Has anyone else noticed this? (I was able to reproduce it with the code posted here by @heikkilamarko as well.)

This is a pretty big problem for us as well, any progress on this issue?

Some additional information:
Azure Functions Core Tools - 2.7.1846
Function Runtime Version - 2.0.12858.0

**Comment by pereiren** [comment]
Hi, happening for me as well.

Using v3-preview 
```  
  <PropertyGroup>
    <TargetFramework>netcoreapp3.0</TargetFramework>
    <AzureFunctionsVersion>v3-preview</AzureFunctionsVersion>
  </PropertyGroup>
```

![image](https://user-images.githubusercontent.com/11976485/69006845-78a1ae80-0935-11ea-9838-c3f073c30c03.png)

`[17/11/2019 11:27:03] Microsoft.Azure.WebJobs.Script.WebHost: Registered factory delegate returns service  is not assignable to ambiently scoped container with {no name}.`


```
[assembly: FunctionsStartup(typeof(UrbanNoise.Importer.Components.Application.Startup))]

namespace UrbanNoise.Importer.Components.Application
{
    public class Startup : FunctionsStartup
    {
        public IConfiguration Configuration { get; }

        public override void Configure(IFunctionsHostBuilder builder)
        {
            var configuration = new ConfigurationBuilder()
                .SetBasePath(Environment.CurrentDirectory)
                .AddJsonFile("local.settings.json", true, true)
                .AddEnvironmentVariables()
                .Build();

            builder.Services.Configure<AppSettings>(configuration);

            builder.Services.AddSingleton<IDocumentClient>(serviceProvider => {
                 DbConnectionStringBuilder cosmosDBConnectionStringBuilder = new DbConnectionStringBuilder
                 {
                     ConnectionString = Environment.GetEnvironmentVariable("Database:ConnectionString")
                 };

                 if (cosmosDBConnectionStringBuilder.TryGetValue("AccountKey", out object accountKey) && cosmosDBConnectionStringBuilder.TryGetValue("AccountEndpoint", out object accountEndpoint))
                 {
                     return new DocumentClient(new Uri(accountEndpoint.ToString()), accountKey.ToString());

                 }

                 return null;
             });

            builder.Services.AddLogging();

            builder.Services.AddSingleton<IGenericComponentsImportService, GenericComponentsImportService>();
            builder.Services.AddSingleton<IGenericComponentRepository, GenericComponentRepository>();
        }
    }
}
```

```
namespace UrbanNoise.Importer.Components.Application
{
    public class Function
    {
        private readonly IGenericComponentsImportService _genericComponentsImportService;

        public Function(IGenericComponentsImportService genericComponentsImportService)
        {
            _genericComponentsImportService = genericComponentsImportService;
        }

        [FunctionName("Function")]
        public async Task Run([TimerTrigger("* * * * *")]TimerInfo myTimer, ILogger log) //("0 * * * *")
        {
            await _genericComponentsImportService.SaveNoiseComponents();
        }
    }
}
```

**Comment by t-l-k** [comment]
@buyckkdelaware during load testing I have found multiple instances of same service type within the same request scope, it occurs frequently for me, at least 1 in 100 requests. I don't know if the _weight_ of the type is relevant, but my `DbContext` subclasses are definitely affected - I get missed saves because entities get written to an incorrect `DbContext`, injected into either repositories, or other infrastructure code. I haven't yet managed to run more than 500 requests due to the DryIOC container scope exploding (which appears to be due to `DefaultHttpClientFactory`, also likely implicated in the `HttpClient` errors you've witnessed).

Detecting these duplicates is quite easy, using an `AsyncLocal` static collection in the affected class's constructor to track when more than 1 distinct hashcode has been found in. DbContexts are not thread safe ofc so can't be used by multiple threads anyway. 

I haven't yet "handled" it, but I'll probably just throw an exception and fail the request thread using hashcodes as the detection basis. Without handling it, my code sometimes doesn't know that it has failed, it just misses data.

**Comment by darrenhull** [comment]
Are there any workarounds for this issue?

**Comment by fabiocav** [comment]
@darrenhull the workarounds will require some code refactoring at the moment. This is being assigned to a sprint for investigation, so we hope to have an update on this issue soon.

**Comment by darrenhull** [comment]
Do you have an update on when you think this may be released? It’s a pretty big blocking issue for my company’s migration which we are pushing further and further away without being able to suggest when we can resume.

**Comment by darrenhull** [comment]
Hi @fabiocav,

I see this has been triaged and had the p2 tag. What exactly does the p2 mean, guessing priority2?

Can you give us an update please?

Thanks

**Comment by fabiocav** [comment]
@darrenhull yes, this is a priority label. The issue has been triaged, but not yet assigned. We'll have the milestone updated once this is assigned to a sprint.

**Comment by daniel-white** [comment]
@fabiocav: this really sounds like a .net core bug/issue. can you confirm? 

i'm running into something similar using a vanilla asp.net core service.

**Comment by bartdk-be** [comment]
@fabiocav : Not to blame anyone, but how can a basic feature like Scoped or not being labeled as a p2 ? This seems like a rather basic functionality ?
Can you point us to workarounds to resolve this issue ?

**Comment by daniel-white** [comment]
This appears to be a problem on at least .NET Core 2.1, but is fixed in 3.x. I've raised it here:  dotnet/runtime#34586

**Comment by bartdk-be** [comment]
@daniel-white 
Any pointer on where it should be fixed in 3.1 ? Because we faced a similar issue in 3.1 (using the functions DI setup). Thanks !

**Comment by daniel-white** [comment]
@bartdk-be: im not sure - in my miminal example at dotnet/runtime#34586 make sure your `Microsoft.Extensions.Http` is the latest version.

**Comment by rodhemphill** [comment]
Hi @t-l-k ... I guess you don't have a work around then? Same significant problem here - fighting with it for months. It seems to happen on all but trivial systems - certainly with three of our serverside blazor systems with between 30 and 50 services. We are thinking of writing a state memory solution - creating a "StateMemoryService" as a singleton and having it register the instances of each other services against the transaction's SessionId. An in-memory register of <sessionId, serviceType, instanceId>.  It's a LOT of effort for what is essentially a bug. @fabiocav @daniel-white

**Comment by melbappdev** [comment]
Hey @fabiocav, if you are up for fixing this I can help. I have three client projects that have the problem and I can easily produce a situation that doesn't get the problem. I can't give you client repos but I have a testing repo that I can share with you. The problem is easy to reproduce and demonstrate in that repo.  This repo includes EF and various service classes, some of which cascade inject other services ... which are the most likely candidates for the problem. Whatever causes it the symptoms are that all "scoped" services get translated to "transient".

**Comment by t-l-k** [comment]
@rodhemphill sticky tape and plasters I'm afraid. Basically failing requests and relying on the runtime to execute the request again with fingers crossed for subsequent attempts. We detect the problem in constructors of classes we know should be scoped, and thus when two instances are created in a logical execution scope (determined by `AsyncLocal`) we throw an exception and log it. Makes a mess of my logs - see https://github.com/Azure/azure-functions-host/issues/4914#issuecomment-601251057. 

It caused quite an embarrasing false start to our UAT phase for sure (deadlines and all), customer was quite alarmed at the amount of failing transactions, and that was as I was chasing it here to get it fixed. I had to refactor in the end, as is the advice per @fabiocav. even changing the order of parameters in constructors reduced the pain! It's still failing on some non-critical code paths, but we can live with that.

Agree- anything non-trivial causes it to fail. We built a polyglot solution, but one component was full DDD stack based on Microsoft's own patterns advice due to the domain complexity. Our simpler Functions Apps (just plain HTTP and read/write storage) work without issue.

We've dropped functions from our list of candidate dotnet technologies for future products & projects in favour of plain ASP.NET Core until this is fixed. Obviously we lose the serverless benefits.

Still, **I managed to get money back from Azure for the development headaches it caused**, as it is a GA product. If you're suffering the same I would recommend you to also raise support requests. Of course, it didn't come close to covering the amount of expense, inconvenience, and damage to our reputation. I had a very professional support experience with Microsoft and their Functions support team were amazing. But they couldn't fix the underlying issue - that's clearly an engineering problem. 

_Perhaps they might treat this more seriously if having to refund more customers via support requests ......... 😉_

**Comment by espray** [comment]
I have had all of my teams stopped all .Net Core AzFunc development, this Scoped DI issue was the final straw for me. If basic .Net Core DI dose not work who knows what other problems may show up at the wrong time, the risk is too great.
  
May be send a message/tweet to Damian Edwards @DamianEdwards and/or Scott Hanselman @shanselman about your AzFunc development experience.

**Comment by AnthonyIacono** [comment]
I think the thing that bothers me about this issue is that it has been around in some form or another for like a year and a half. The original bug was a two line fix, and I imagine the issue we are all facing currently is similar. Can we please get an update from engineering?

**Comment by darrenhull** [comment]
We too have had to stop using @azfunc as the scoping issue makes them unusable for our solution. This is very frustrating for a team that have embraced Azure functions from day one. I’ve been very disappointed with how long this issue has taken to be resolved @fabiocav , oh wait it’s still active! We too are seeking money back on our subscriptions.

**Comment by melbappdev** [comment]
I have wasted about 4 days solid on this issue and with impending delivery dates and (soon to be) an unhappy client ... I am glad you responded @t-l-k as you've saved me a lot more wasted time. Yes I agree @darrenhull that it sounds like it has taken too long to fix, but a more important question is: what has it cost? How many people like me have fallen into this hole? @fabiocav

**Comment by Arash-Sabet** [comment]
This issue is a huge blocker and addressing it has to be expedited. If there's a P1 label indicating a higher priority than P2, that one should be applied instead of P2. We cannot proceed with our project gracefully without a fix! @fabiocav

**Comment by fabiocav** [comment]
Assigning this to the next sprint. @Arash-Sabet, if you can open an issue detailing your scenario, we might be able to assist with a workaround until this is resolved.

**Comment by fabiocav** [comment]
> The original bug was a two line fix, and I imagine the issue we are all facing currently is similar. 

@anthonychu thank you for the patience, can you please share more details on what fix you're referring to?

**Comment by Arash-Sabet** [comment]
@fabiocav thanks for your reply. I will try to get you a bare-minimum project tomorrow. The only workaround that I could think of was to avoid the "scoped object" dependency injection and instead pass the "scoped" object's instance by adding an interface property to the consumer/service object. I was just wondering, what workaround you may be thinking of so that I could think about it ahead of time? Thanks.

**Comment by vitalybibikov** [comment]
Impossible to implement Domain Events with Azure functions without this functionality properly...
Can't believe it does not work. Still....

I'm asking to change the priority, as this is a blocker.

Thanks

**Comment by vitalybibikov** [comment]
Hi @fabiocav 
Any news on this?

Can I somehow assist to speed this one up?

Thanks

**Comment by vitalybibikov** [comment]
As my solution is heavily using DDD, it was not possible for me to ignore absence of this core functionality. 

I am using Bindings a lot, so moving to App Service is also not an option at the moment.
For anybody who is planning to use AF in prod, I'm strongly advising not to do so, as they are definitely not in Production Ready state for anything except for "web hooks".

The solution for now is:

To emulate scoped by using AddSingleton and AddTransient, so it will be possible to store all the connections in a Singleton dictionary and grab them from their when necessary.

Connection Key: FullTypeName + Request HashCode (or anything you want)

In startup it looks like this:

`            services.AddScopedContext<ITenantDbContext, TenantDbContext>();`

Some boilerplate is required in order to dispose everything on exit.
My solution is to use DisposeContextFilter class that encapsulates all the disposal logic.

As filters in AF are not in a ready-to-use state also, it looks ugly. Though it works.
I can share the code if necessary.

**Comment by Boaz101** [comment]
I am also experiencing issues with an Azure function which uses scoped services combined with AddHttpClient. If I comment the AddHttpClient from startup it works correctly. I am moving the HttpClient out of dependency injection, however that has potential pitfalls, too. Hopefully, this is fixed soon because this is causing us a lot of headaches.

**Comment by roszell** [comment]
@fabiocav any news on this? I see it was assigned to Sprint 75 that was marked closed, but this was the only issue not resolved as part of that. Is this actively being worked on still?

**Comment by Schlebusch** [comment]
Sooooo ... any news? Glad to find this issue early enough to stop using scoped dependencies in my Azure Functions project.
It is frustrating to read that this issue is open for 9 months now and nothing indicates that it will be solved at all ...
MSFT should create a list of features in their SDK's that are not working.

**Comment by APIWT** [comment]
@fabiocav Hope you are having a nice day :) Any updates on this?

**Comment by t-l-k** [comment]
@fabiocav please can you advise as to any updates for this issue? We've some gaps opened up in our roadmap in August in which we're looking forward to paying down some technical debt, and revisiting this issue is on our tracker alongside migrating to dotnet 3+.

Is this recognised as a potential https://github.com/dadhi/DryIoc issue? Is there a correlating ticket with the DryIOC project? 

Thanks

**Comment by espray** [comment]
@jeffhollan @brettsam any update on this issue?

**Comment by alexanderbikk** [comment]
hi guys
I'm using .net core 3.1 and  looks like Scoped services work well but not for the "cold start".
 So, when I run the function app either on local or Azure env and the function does the "cold start", my scoped services will be injected more than once per request. When I do the requests after "cold start" seems like all good.

So, for example EF DbContext will be created more that once per request for the "cold start". But for all next requests it will be only the one instance per request.
I'm wondering does it the same issue or I'm doing something wrong?

btw same for my custom services. I create the UoW which includes the EF DbContext, and registered it as Scoped. And again got the same issue.

Thanks

**Comment by rodhemphill** [comment]
Hi @alexanderbikk. One thing to check is whether you have a "prerender" process running that creates the initial instance. This is a known problem with net core Blazor apps for example.

**Comment by alexanderbikk** [comment]
Hi @rodhemphill I'm not sure that I get you.

As I said the Scoped lifitime works sometimes, but sometimes I get different instances per request. I can reproduce it on "cold start". Then everything works fine. After some idle the function does a "cold start" again and I again get this issue.

Thanks

**Comment by brettsam** [comment]
@alexanderbikk -- would you be able to throw a simplified repro up somewhere (in a github repo or even just a gist)? That'd help get this set up...

**Comment by alexanderbikk** [comment]
hi @brettsam 

I believe I have not so much time for it. But I can describe the details and provide some code.
You can try the following scenario:
- Create the Azure function v3 with latest sdk version.
- Create  two or more "Services"  and UoW which can be configured as Transient  in function Startup
- Add EF core DbContext which will be configures as Scoped by default. I used separate project with EF Core and just configuring the DbContext in function Startup. 
- Use created DbContext  in "Services" and UoW  (DI in constructors)
- Call UoW  and "Services" in az function to check the injected DbContext   instance
- Also you can add some kind of logic in Services and using UoW  to save changes. But it will be enough to check the injected DbContext  InstanceId

The expected behavior is that the DbContext   instance should be the same per one function call
But in my case sometimes(I reproduced it on "cold start") the "Context" instance is different for different Services, so I got some kind of Transient behavior.

Here is possible  example of Startup.cs and Services

    public class Startup : FunctionsStartup
    {
        public IConfigurationBuilder ConfigurationBuilder { get; }

        public override void Configure(IFunctionsHostBuilder builder)
        {
            builder.Services.AddLogging();

            builder.Services.AddHttpClient();

            builder
                .Services
                .AddOptions<ApplicationSettings>()
                .Configure<IConfiguration>((settings, configuration) => { configuration.Bind(settings); });

            ApplicationSettings config = builder.Services.BuildServiceProvider().GetRequiredService<IOptions<ApplicationSettings>>().Value;
            builder.Services.AddDbContext<DatabaseContext>(options => options.UseSqlServer(config.WebSql.ConnectionString));

            builder.Services.AddTransient<IService1, Service1>();
            builder.Services.AddTransient<IService2, Service2>();
            builder.Services.AddTransient<IUnitOfWork, UnitOfWork>();

        }
    }

    public class Service1 : IService1
    {
        private readonly DatabaseContext dbContext;

        public Service1(DatabaseContext dbContext)
        {
            this.dbContext = dbContext;
        }
    }

    public interface IService1
    {
    }

    public class Service2 : IService2
    {
        private readonly DatabaseContext dbContext;

        public Service2(DatabaseContext dbContext)
        {
            this.dbContext = dbContext;
        }
    }

    public interface IService2
    {
    }

    public class UnitOfWork : IUnitOfWork
    {
        public DatabaseContext DbContext { get; }

        public UnitOfWork(DatabaseContext dbContext)
        {
            DbContext = dbContext;
        }

        public void BeginTransaction()
        {
            DbContext.Database.BeginTransaction();
        }

        public async Task BeginTransactionAsync()
        {
            await DbContext.Database.BeginTransactionAsync();
        }

        public void Commit()
        {
            DbContext.Database.CommitTransaction();
        }

        public void Rollback()
        {
            if (DbContext.Database.CurrentTransaction != null)
            {
                DbContext.Database.RollbackTransaction();
            }
        }

        public void SaveChanges()
        {
            DbContext.SaveChanges();
        }

    } 

UPD: And one important note that I'm using bot UoW and Services not directly in function but inside the other class, let's call it Manager. Which injected and called from azure function.

    public class Manager : IManager 
    {
        private readonly IService1 _service1;
        private readonly IUnitOfWork _unitOfWork;

        public StateMachineActionsInvoker(IService1 service1,
            IUnitOfWork unitOfWork)
        {
            _unitOfWork = unitOfWork;         
            _service1 = service1;
        }
        public void Call()
        {           
           //do something         
         }
    }

    public class TestFunction
    {
        private readonly IManager _manager;

        public TestFunction(IManager manager)
        {
            _manager= manager;
        }

        [FunctionName("TestFunction")]
        public async Task<IActionResult> Run(
            [HttpTrigger(AuthorizationLevel.Function, "get", "post", Route = null)] HttpRequest request,
            ILogger log)
        {
            _manager.Call()
        }
    }
   
UPD2: @brettsam I have tried to reproduce it and  no luck

 I think I found the issue. At least I found the case when it happens. I called the dbContext.SaveChanges() inside my services and outside of services I called UoW, to use the transaction. In this case the injected dbContext (InstanceId) was different for UoW and my Service. But again it was only on "cold start" really strange behavior.

I have refactored my code, and called dbContext.SaveChanges() outside of the services from my UoW and I can't reproduce the issue. Seems like all good and nothing to check

Thanks a lot.

**Comment by pragnagopa** [comment]
@brettsam - Any updates here? Should we move this next sprint?

**Comment by brettsam** [comment]
I haven't been able to investigate yet -- but I've moved it to the next sprint so it comes back on @fabiocav's radar.

**Comment by fsalmeida** [comment]
I've been facing the same problem and I've found a workaround.
While I was testing some scenarios, I found out that the problem only happens on classes that expect the HttpClient as DependencyInjection.
For example:

`
public class SomeClass
{
    private readonly MyHttpClient myHttpClient;
    private readonly MyScopedDependency myScopedDependency;

    public SomeClass(MyHttpClient myHttpClient, MyScopedDependency myScopedDependency)
    {
        this.myHttpClient = myHttpClient;
        this.myScopedDependency = myScopedDependency;
    }

    public void SomeMethod()
    {
        myHttpClient.DoSomething();
    }
}
`

In this scenario, the "MyScopedDependency" would not be resolved as Scoped, but if I change my code to something like this:

`
public class ClassToHoldHttpClient
{
    public MyHttpClient MyHttpClient { get; }

    public ClassToHoldHttpClient(MyHttpClient myHttpClient) => MyHttpClient = myHttpClient;
}

public class SomeClass
{
    private readonly ClassToHoldHttpClient classToHoldHttpClient;
    private readonly MyScopedDependency myScopedDependency;

    public SomeClass(ClassToHoldHttpClient classToHoldHttpClient, MyScopedDependency myScopedDependency)
    {
        this.classToHoldHttpClient = classToHoldHttpClient;
        this.myScopedDependency = myScopedDependency;
    }

    public void SomeMethod()
    {
        classToHoldHttpClient.MyHttpClient.DoSomething();
    }
}
`

With this code, "MyScopedDependency" is now beeing resolved as Scoped again.

Is it a great solution? Of course not, but as this problem is not solved for about to 1 year, maybe this workaround may save some people.

**Comment by GuhBigardi** [comment]
> I've been facing the same problem and I've found a workaround.
> While I was testing some scenarios, I found out that the problem only happens on classes that expect the HttpClient as DependencyInjection.
> For example:
> 
> `
> public class SomeClass
> {
> private readonly MyHttpClient myHttpClient;
> private readonly MyScopedDependency myScopedDependency;
> 
> ```
> public SomeClass(MyHttpClient myHttpClient, MyScopedDependency myScopedDependency)
> {
>     this.myHttpClient = myHttpClient;
>     this.myScopedDependency = myScopedDependency;
> }
> 
> public void SomeMethod()
> {
>     myHttpClient.DoSomething();
> }
> ```
> 
> }
> `
> 
> In this scenario, the "MyScopedDependency" would not be resolved as Scoped, but if I change my code to something like this:
> 
> `
> public class ClassToHoldHttpClient
> {
> public MyHttpClient MyHttpClient { get; }
> 
> ```
> public ClassToHoldHttpClient(MyHttpClient myHttpClient) => MyHttpClient = myHttpClient;
> ```
> 
> }
> 
> public class SomeClass
> {
> private readonly ClassToHoldHttpClient classToHoldHttpClient;
> private readonly MyScopedDependency myScopedDependency;
> 
> ```
> public SomeClass(ClassToHoldHttpClient classToHoldHttpClient, MyScopedDependency myScopedDependency)
> {
>     this.classToHoldHttpClient = classToHoldHttpClient;
>     this.myScopedDependency = myScopedDependency;
> }
> 
> public void SomeMethod()
> {
>     classToHoldHttpClient.MyHttpClient.DoSomething();
> }
> ```
> 
> }
> `
> 
> With this code, "MyScopedDependency" is now beeing resolved as Scoped again.
> 
> Is it a great solution? Of course not, but as this problem is not solved for about to 1 year, maybe this workaround may save some people.

\o/ It worked for me, thank u @fsalmeida, but I hope that fix coming soon :)

**Comment by espray** [comment]
@brettsam @fabiocav @jeffhollan @shanselman @DamianEdwards this was part of Spring 84, but was not completed. Now its not part of Sprint 85! **WTF!** Its coming up on 1yr for this **P2** issue. Not building a lot of confidence in the community this way.

**Comment by vitalybibikov** [comment]
@espray I guess that's some kind of development, when you can just drop things out of a scope forever and ever.
As I can see 2 or 3 things in each of their sprints, that are not closed and just hang there without being tackled.

I've been waiting almost for a year for this thing to be closed. 
But I guess something like Scoped is not a priority, so I've just created my own Scoped extension out of Singleton and Transient for one of my customers. 

Now, as an architect in my organization I'm advising client's not to use Azure Functions, as they are still not ready for production use.

**Comment by espray** [comment]
@vitalybibikov same situation here. Actively not recommend AzFunc as well. Mainly because of **HOW** this issue has been handled.

**Comment by fabiocav** [comment]
Apologies for the delay on this. We have been, we have a relatively small team, and principally during the current circumstances, it has been challenging to address some of the issues in our sprints, and this was unfortunately, one of them.

I've been working with @brettsam and @anthonychu to resume the work on this. I also apologize for the lack of communication here. We'll keep this issue updated as we move forward.

Thank you for the patience.,

**Comment by fabiocav** [comment]
Providing an update here at the end of the week. @brettsam and I have made some progress on the different scenarios reported, testing and investigation, hopefully we'll have something ready before the sprint is over, but will continue to provide updates here for awareness.

**Comment by APIWT** [comment]
@brettsam Glad to see the fix was very simple. Would you be able to provide a timeline on deployment?

**Comment by brettsam** [comment]
The code change was small, but changes like this can be pretty impactful, which is why we decided to hide it behind a feature flag and get that change out ASAP. We wanted to let everyone try it and let us know if they run into issues while we continue testing scenarios. If all goes smoothly, we will make this the default everywhere in a future release.

As for the timeline -- it's currently in validation and the deployment is starting this week. We don't like giving exact timetables (hotfixes, unrelated incidents, etc can delay a deployment)... but it generally takes 6-ish business days to roll out worldwide once it has started.

**Comment by APIWT** [comment]
@brettsam Sounds good! Is there an easy way to be notified when the runtime on our functions app resource has support for the feature flag?

**Comment by darrenhull** [comment]
Thanks for fixing the issue and for the feedback @brettsam and @fabiocav. I too would like to know when it is available.

**Comment by t-l-k** [comment]
A quick GitHub search: https://github.com/Azure/azure-functions-host/wiki/Configuration-Settings - 

> `AzureWebJobsFeatureFlags`: Set to a comma delimited list of beta features to enable. Beta features enabled by these flags are not production ready, but can be enabled for experimental use before they go live.

I'm putting this in the AppSettings of my test deployments _posthaste_ and grabbing the popcorn:

`AzureWebJobsFeatureFlags` = `EnableEnhancedScopes`

Meanwhile I'll be watching https://github.com/Azure/azure-functions-host/releases and https://github.com/Azure/azure-functions-core-tools/releases like a hawk waiting for this to drop.

**Comment by fabiocav** [comment]
The validation phase has been completed and the release is starting to rollout globally. We don't usually publish the Core Tools until after the deployment, but we're making some changes to the process and will see if we can get this up ASAP so you can test locally as well.

Again, thanks for the patience with this!

**Comment by fabiocav** [comment]
We'll actually keep this issue open (the original intention was to keep it open) so we can continue to use this for feedback/validation.

**Comment by ebwinters** [comment]
Guys I have the solution after hours of investigation:
Instead of `async void` for function return type, do `async Task`. Not sure why this fixes it, any explanation would be great. @fabiocav

**Comment by jeffhollan** [comment]
In C# you should only use `async void` when you have to (event handlers, which functions are not in that sense), so likely due to that https://stackoverflow.com/questions/13636648/wait-for-a-void-async-method

**Comment by ishepherd** [comment]
> Not sure why this fixes it, any explanation would be great.

@ebwinters your function would return when it hits the first `await`. I suppose the functions host has no way of telling there is still work pending, because you did not return a `Task` for it to wait on.

> The async void case is a “fire and forget”: You start the task chain, but you don’t care about when it’s finished. When the function returns, all you know is that everything up to the first await has executed. Everything after the first await will run at some unspecified point in the future that you have no access to.
> https://devblogs.microsoft.com/oldnewthing/20170720-00/?p=96655

**Comment by ebwinters** [comment]
Good to know

**Comment by t-l-k** [comment]
This is now working great for me, my failed request rates and exception rates have basically dropped to zero, and applications are working better than ever. This brings Azure Functions back to the table for projects in flight!!!

p.s. I skipped local validation using the `func.exe` tools and just went for the docker image `mcr.microsoft.com/azure-functions/dotnet:3.0.14785`, which worked great as well.

**Comment by brettsam** [comment]
@t-l-k -- to be clear, you're running with the new feature flag?

**Comment by t-l-k** [comment]
@brettsam indeed! Unless my eyes deceive me!

![image](https://user-images.githubusercontent.com/15836198/97586848-a4a52b00-19f2-11eb-8c7e-dece280de5b5.png)

**Comment by fabiocav** [comment]
@t-l-k great! Thanks for providing feedback on this! And again, thank you for the patience with this! If you find any other issues while enabling this toggle, please let us know as we'll be making this the default once we have some good validation.

For others who were experiencing the same issue, please provide feedback if possible.

> Guys I have the solution after hours of investigation:
> Instead of async void for function return type, do async Task. Not sure why this fixes it, any explanation would be great. 

@ebwinters the information provided by @ishepherd is correct, the pattern you were using will trigger the behavior you're describing and other potential side effects as well (e.g. early termination of your function if the host is deallocated). Without a `Task` that represents the work being executed, the calling code doesn't have any information about when that work completes and the invocation will be completed as soon as the method returns. By the way, this is not unique to Azure Functions, but applies to .NET in general. We do have checks and warnings in the runtime (if you were to look at you app diagnostics, you'd see a warning about `async void` usage) to make you aware of that.

**Comment by ebwinters** [comment]
@fabiocav that makes sense, thanks for explaining. I saw some code snippets in this thread with a void return type, so this may still help them (assuming warnings are being suppressed)

**Comment by ghost** [comment]
@fabiocav Applying the feature flag helped in our case as well.

Our story:
Last month, we noticed some odd behavior regarding scoped services. In our case, a scoped service was responsible for setting up one database connection per request if necessary, and disposing it once the scoped service was disposed. However, after a longer period of time without restarts, the function app ran out of ports. After some (long) investigation we realized that scoped services were sometimes not disposed correctly.
Then I came across this issue 2 weeks ago and we applied the feature flag to the function app that showed this behavior. After that we let it run for a while, and it seems that we no longer leak connections, which suggests that the scoped service is being disposed of properly now.

After 2 months since the last response, is there any idea of when this will be available by default without having to add the feature flag?

**Comment by brettsam** [comment]
Thanks for providing that data, @b10-dslappendel! This is very useful.

We'll discuss this and see if we have enough usage and data to make it the default.

**Comment by simongullberg** [comment]
I can tell that adding "AzureWebJobsFeatureFlags": "EnableEnhancedScopes" solved the problem for us when injecting a scoped service.

We were injecting a scoped EF DbContext and this context was injected in two transient services (let's call those Service1 and Service2). Service1 fetched some entity from the database with the injected DbContext and then sent the entity to Service2 where the entity was modified and a call to SaveChangesAsync() was made on the injected DbContext in Service2. The update did not work because there was no change when looking in the database. I suspected that something was wrong with the DbContext so I added an instance variable of type Guid and called Guid.NewGuid. I then expected to see the same Guid on the DbContext injected in Service1 and Service2 but they were different. After adding the feature flag EnableEnhancedScopes the Guids were the same in both Service1 and Service2 and the entity was updated in the database.

**Comment by danielreuterwall** [comment]
We've had similar issues. Change tracking didn't seem to work between EF DbContext injected directly in function code and EF DbContext injected in scoped services. Using the feature flag did solve that problem for us which is great.

However, in a service added using `AddHttpClient` in our `Startup.cs` we injected `IHostEnvironment environment` to determine if we were in a production environment or not. After the feature flag was set `IHostEnvironment environment` was no longer available. We've solved it using a regular configuration variable, but the reason for why it's no longer available is unclear. We also tried to revert by removing the feature flag configuration but is seems like once it's set it's there for good.

So two questions:
Any reason or explanation why `IHostEnvironment environment` is not longer available?
Is it possible to remove the feature flag?

**Comment by VaclavK** [comment]
stumbled upon this while investigating issue with incomplete ef core (3) persistance..

our pattern is that we have multiple repos supposedly using (scoped) DbContext and we call save changes on one of them - this generally works (is meant to) but in one function app we have http client injected into DI

we started seeing comply partial data persistence (from the repo that was saved, not from the others) which this issue explains

this feature flag - is this PROD ready - assuming this is considered beta flag so it would not be recommended for production usage??? @fabiocav 

reproducible in azure (v3 functions) and locally with azure functions CLI - 3.23.0

interestingly enough - we were trying to understand why we would not see this issue consistently/with high enough frequency - our (affected) function is triggered in parallel by SB trigger and behaviourally it seems not to be reproducible with message pump invoking function in parallel (in default config up to 16 messages)

**Comment by t-l-k** [comment]
@VaclavK I wouldn't consider Azure Functions "production ready" when running _without_ this flag TBH, especially if you're running a design with a properly encapsulated domain model with plenty of service abstractions.

@brettsam Is it enabled by default yet?

**Comment by fabiocav** [comment]
@t-l-k I'm sending the changes to make this the default in the current sprint. Next release will have the new behavior.

**Comment by VaclavK** [comment]
@fabiocaFabie just so I understand this - This will be the default behaviour without need to configure anything in?

How do I know that when that release will be the release running in azure for my functions?

**Comment by swettstein** [comment]
@fabiocav there was a release 2 days ago (3.0.15733). can you confirm this was part of that release?

**Comment by fabiocav** [comment]
@swettstein are you referring to the change in defaults? If so, the change was not made because of concerns with a few other items landing right before //build and some announcements.

My goal is to have the defaults change with the following release.

**Comment by mintyfusion** [comment]
We ran into the same issue, but adding AzureWebJobsFeatureFlags = EnableEnhancedScopes is making things worse, where all Scoped objects behave as Transient.

We are using Azure Functions 3.x and I have tried both Windows and Linux and result is exactly the same.

**Comment by fabiocav** [comment]
@mintyfusion can you please open an issue with the details of what you're seeing a a repro? What you're describing above leads me to believe something is incorrect with the setup and it would be good to focus on that.

**Comment by swettstein** [comment]
@fabiocav any update on releasing "EnableEnhancedScopes" so we don't need the feature flag anymore?

**Comment by Socolin** [comment]
Hello,

~I also have this issue, but `EnableEnhancedScopes` dont fix the problem, is this supposed to works with the new azure functions (.net 5 + runtime: dotnet-isolated)~

Edit:
I finally found the problem, there was a service registerd as Singleton and depending on a Scoped dependency (I was expecting an exception instead of a silent `new()`)

**Comment by qcnguyen** [comment]
any updates on this ? team ? @fabiocav

**Comment by ghobakhloum** [comment]
I had the same issue, and resolved it by using autofac in azure function v5

**Comment by swettstein** [comment]
@fabiocav any update please?

**Comment by Schuer84** [comment]
Any updates on this one? We registered our DbContext as scoped, but we noticed, if we start multiple http requests, the context gets shared between the different requests, which leads to threading issues: 

{
  type: 'NotSupportedException',
  message: "A second operation started on this context before a previous asynchronous operation completed. Use 'await' to ensure that any asynchronous operations have completed before calling another method on this context. Any instance members are not guaranteed to be thread safe."
}

Any feedback on how to fix this would be greatly appreciated. 
PS: We enabled the flag  "AzureWebJobsFeatureFlags": "EnableEnhancedScopes",

**Comment by fabiocav** [comment]
@Schuer84 can you please open a separate issue with those details so we can take a closer look at your scenario? We have no evidence that with the changes made behind the flag still have issues with scoped services, so it would be good to rule other things out. 

Thanks!

**Comment by Gevil** [comment]
Alright some new info on this in our functions we are also using HttpClients with DI with AddHttpClient.

After enabling EnableEnhancedScopes we are seeing fewer errors related to object disposed exceptions from DryIoc related to our DbContexts when locally debugging.

As soon as we deploy our Function to Azure and run tests on it everything falls apart and almost every function trigger fails.

`{
  "Error": 39,
  "StackTrace": "   at DryIoc.Throw.It(Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in D:\\a\\1\\s\\src\\WebJobs.Script.WebHost\\DependencyInjection\\DryIoc\\Container.cs:line 9000\r\n   at DryIoc.Container.ThrowIfContainerDisposed() in D:\\a\\1\\s\\src\\WebJobs.Script.WebHost\\DependencyInjection\\DryIoc\\Container.cs:line 412\r\n   at DryIoc.Container.ResolveAndCacheDefaultFactoryDelegate(Type serviceType, IfUnresolved ifUnresolved) in D:\\a\\1\\s\\src\\WebJobs.Script.WebHost\\DependencyInjection\\DryIoc\\Container.cs:line 201\r\n   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in D:\\a\\1\\s\\src\\WebJobs.Script.WebHost\\DependencyInjection\\DryIoc\\Container.cs:line 196\r\n   at Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection.ScopedServiceProvider.GetService(Type serviceType) in D:\\a\\1\\s\\src\\WebJobs.Script.WebHost\\DependencyInjection\\ScopedServiceProvider.cs:line 25\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)\r\n   at Microsoft.EntityFrameworkCore.Internal.ScopedLoggerFactory.Create(IServiceProvider internalServiceProvider, IDbContextOptions contextOptions)\r\n   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_0(IServiceProvider p)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)\r\n   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()\r\n   at Microsoft.EntityFrameworkCore.DbContext.get_Model()\r\n   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityType()\r\n   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.get_EntityQueryable()\r\n   at Microsoft.EntityFrameworkCore.Internal.InternalDbSet`1.System.Linq.IQueryable.get_Provider()\r\n   at Microsoft.EntityFrameworkCore.EntityFrameworkQueryableExtensions.Include[TEntity,TProperty](IQueryable`1 source, Expression`1 navigationPropertyPath)\r\n   at Gsop.Cpo.ChargeLocationService.Business.Services.OpeningTimesService.GetRegularOpeningHoursPeriodsAsync(String chargeStationId, DateTime utcNow) in D:\\a\\1\\s\\Gsop.Cpo.ChargeLocationService.Business\\Services\\OpeningTimesService.cs:line 348\r\n   at Gsop.Cpo.ChargeLocationService.Business.Services.OpeningTimesService.GetAllPeriodsForAllChargeStations(String chargeStationId, DateTime utcNow) in D:\\a\\1\\s\\Gsop.Cpo.ChargeLocationService.Business\\Services\\OpeningTimesService.cs:line 182\r\n   at Gsop.Cpo.ChargeLocationService.Business.Services.OpeningTimesService.ProcessOpeningHoursPeriods(String chargeStationId, Nullable`1 utcNowOverride) in D:\\a\\1\\s\\Gsop.Cpo.ChargeLocationService.Business\\Services\\OpeningTimesService.cs:line 35\r\n   at Gsop.Cpo.ChargeLocationService.Functions.FunctionsTimerTriggers.CalculateOpeningTimes(TimerInfo myTimer) in D:\\a\\1\\s\\Gsop.Cpo.ChargeLocationService.Functions\\Functions.Timer.cs:line 36",
  "Message": "Container is disposed and should not be used: Container is disposed.\r\nYou may include Dispose stack-trace into the message via:\r\ncontainer.With(rules => rules.WithCaptureContainerDisposeStackTrace())",
  "Data": {},
  "Source": "Microsoft.Azure.WebJobs.Script.WebHost",
  "HResult": -2146233079
}`


The function Version is V3 we're running .net core 3.1 and ef core 3.1 version.


This is ridiculous that we can't use proper DI and HttpClient injection with functions runtime and packages that are in production.

You need to provide at least a workaround otherwise there's no point in using azure functions whatsoever.

Without the Feature flag we can at least get our functions running and only encountering the issue rarely.

**Comment by thomasvdb** [comment]
@Gevil The workaround we found when using HttpClients is by adding an extra abstraction layer and don't inject an HttpClient (or HttpClientFactory) directly. This is now running for about 1 year in production without any issues.

Example:
```
public class FunctionWorkaround
{
        private readonly IHttpClientFactory httpClientFactory;

        public FunctionWorkAround(IHttpClientFactory httpClientFactory)
        {
            this.httpClientFactory = httpClientFactory;
        }
        
        public HttpClient CreateClient() => httpClientFactory.CreateClient(...);
}
```

**Comment by eluchsinger** [comment]
@thomasvdb so what do you inject into the Function then? In the example you are injecting an `IHttpClientFactory`. Do you inject the `FunctionWorkaround`?

**Comment by eluchsinger** [comment]
@fabiocav @brettsam This issue looks like related to my new issue reported. I don't think that it's the same issue though so I wouldn't merge them.

https://github.com/Azure/azure-functions-dotnet-worker/issues/628

**Comment by VaclavK** [comment]
Our strategy so far dealing with this


in our case  (given this is your vanilla ef.core n-tier app with db store, this may be applicable for other folks) was that due to separate db context (and its associated changed entities), data was not being saved as save changes was invoked on the primary db context ,not the secondary

we have moved all functions that talk to Http (fortunately that is few integration points in our apps) to a separate fn app... these are woken by commands and do the http dance and persist whatever they retrieve etc into db store, for many of the data services we have added an optional param to send a repository object to interact with DB - so as to skip the one injected by DI (as that is tied to a separate db context due to this bug)

this way we gain additional runtime isolation - not to frown upon - all tied to the same fn app plan so the overhead is not huge - and the code change associated is easy to dissolve once this is resolved (and we could have done feature flag)

**Comment by thomasvdb** [comment]
@eluchsinger Indeed, I inject the `FunctionWorkaround` (it has an interface as well in our case, left it out in the example).

**Comment by swettstein** [comment]
> @swettstein are you referring to the change in defaults? If so, the change was not made because of concerns with a few other items landing right before //build and some announcements.
> 
> My goal is to have the defaults change with the following release.

@fabiocav  any update on making this feature flag default? i haven't seen/heard anything. our team has been tracking this issue for a long time and we have to add this feature flag to each app we make so would be nice that it gets moved to default.

**Comment by swettstein** [comment]
> @fabiocav any update on making this feature flag default? i haven't seen/heard anything. our team has been tracking this issue for a long time and we have to add this feature flag to each app we make so would be nice that it gets moved to default.

in case anyone else is looking for the answer, i found it here: https://github.com/Azure/azure-functions-host/pull/7669

seems that it will only be removed in v4, not v3

**Comment by soloham** [comment]
I've just experienced this issue when using middlewares (.NET 6, isolated), I have got an auditing scope defined in the middleware before the function call where I resolve a (scoped) EF DB context here and would like to call SaveChanges on this DB context at the end of the request pipeline, but it seems this DB context is different to the one that is resolved in the function, so doing save changes at the end does nothing as the changes never got tracked on this DB context.

Please let me know if this is a bug or expected behavior for the host and if expected how can I achieve the above goal. Thanks!

@fabiocav @brettsam FYI

**Comment by brettsam** [comment]
@SoloHam -- can you open an issue in the https://github.com/Azure/azure-functions-dotnet-worker/issues repo? Please include some example code to show us how things are set up in your scenario.

This issue is related to the in-proc DI setup, which is different from that in the isolated worker.

**Comment by soloham** [comment]
It appears my issue is already answered ([#811](https://github.com/Azure/azure-functions-dotnet-worker/issues/811)) in the appropriate repo. Thanks @brettsam!

**Comment by jonmeyerson** [comment]
I ran into a similar issue in 3.1, I had a ScopedServiceProvider and when HttpClientFactory.CreateClient was called inside it, it caused the next time I resolved a interface some of the instances to be lost. After looking at the source code for I did get it to work setting [SupressHandlerScope](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.http.httpclientfactoryoptions.suppresshandlerscope?view=dotnet-plat-ext-6.0&viewFallbackFrom=netcore-3.1) to true. Not sure what side effects this has. 
its used to control this logic in the HttpClientFactory - [DefaultHttpClientFactory](https://github.com/dotnet/runtime/blob/3ae87395f638a533f37b8e3385f6d3f199a72f4f/src/libraries/Microsoft.Extensions.Http/src/DefaultHttpClientFactory.cs#L122)

**Comment by jorge-duarte** [comment]
I confirm that the issue persists even in version 4 of the Azure Functions runtime. Worst of all, I don't use the IHttpClientFactory interface and yet AddScoped always generates new instances on each constructor injection.

**Comment by evaldas-raisutis** [comment]
Found this thread after banging by head against he wall for 3 days.

We had a transient class A which depended on a scoped dbcontext and another scoped service B with httpclient added through httpclient nuget package and dbcontext

I could reproduce an issue where the first time service A was resolved through a scoped service provider, two instances of dbcontext were created. Service A and Service B receives separate instances of db context. 

Service B modified entities through dbcontext and save changes was called on service A. Because there were two instances of dbcontext and each service wrongly had its own instance - nothing was saved to the database.

Subsequent resolutions of service A did not produce the same behavior and both service A and service B received the same instance of dbcontext.

Enabling the feature flag mentioned earlier helped, but I am yet to test it on production.

**Comment by campelo** [comment]
He everybody. 

I see this issue has been opened for almost 3 years. I'm using all the most recent versions and I'm continue facing the same issue.
Could we have a permanent solution without workarounds?

**Comment by brettsam** [comment]
@campelo -- are you on v4? Can you please provide an example of what you're seeing in a new issue and tag me and @fabiocav in it? There are a lot of mixed scenarios in this issue and it's getting hard to follow.

**Comment by brettsam** [comment]
> I confirm that the issue persists even in version 4 of the Azure Functions runtime. Worst of all, I don't use the IHttpClientFactory interface and yet AddScoped always generates new instances on each constructor injection.

@jorge-duarte -- Without seeing your exact repro, I believe this is working as expected. Each function invocation is its own scope, which mean it gets its own instance of scoped services. Similar to an http request in ASP.NET. Could you also open a separate issue and tag me and @fabiocav in it so we can take a look at your exact scenario?

**Comment by Assassinbeast** [comment]
I also have Scoped lifetime problem with Isolated runtime.

It looks like there are two scoped lifetimes for one http request.
One for the "Middlewares" and one for the "Function".

Here is my code:

**Startup.cs**
```
var host = new HostBuilder()
    .ConfigureFunctionsWorkerDefaults((IFunctionsWorkerApplicationBuilder configure) =>
    {
        configure.UseMiddleware<MyMiddleware>();
        configure.UseMiddleware<MyMiddleware2>();
    })
    .ConfigureServices(services =>
    {
        services.AddScoped<FooService>();
        services.AddScoped<BarService>();
    })
    .Build();
host.Run();
```

**Function1.cs**
```
namespace FunctionApp1
{
    public class Function1
    {
        private readonly FooService _fooService;
        private readonly BarService _barService;

        public Function1(FooService fooService, BarService barService)
        {
            _fooService = fooService;
            _barService = barService;
        }

        [Function("Function1")]
        public HttpResponseData Run([HttpTrigger(AuthorizationLevel.Function, "get", "post")] HttpRequestData req)
        {
            _fooService.LogFooId();
            _barService.Start();
            return req.CreateResponse(HttpStatusCode.OK);
        }
    }

    public class MyMiddleware : IFunctionsWorkerMiddleware
    {
        private readonly FooService _fooService;

        public MyMiddleware(FooService fooService)
        {
            _fooService = fooService;
        }

        public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
        {
            _fooService.LogFooId();
            await next(context);
        }
    }
    public class MyMiddleware2 : IFunctionsWorkerMiddleware
    {
        private readonly FooService _fooService;

        public MyMiddleware2(FooService fooService)
        {
            _fooService = fooService;
        }

        public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
        {
            _fooService.LogFooId();
            await next(context);
        }
    }

    public class FooService
    {
        public Guid Id { get; set; } = Guid.NewGuid();
        private readonly ILogger _logger;

        public FooService(ILogger<FooService> logger) 
        {
            _logger = logger;
        }

        public void LogFooId()
        {
            _logger.LogInformation($"FooId: {Id}");
        }
    }

    public class BarService
    {
        private readonly FooService _fooService;

        public BarService(FooService fooService)
        {
            _fooService = fooService;
        }

        public void Start()
        {
            _fooService.LogFooId();
        }
    }
}
```

When i call "http://localhost:7071/api/Function1", then it will log this out:

> FooId: 2514bc34-f5d6-4833-b14b-5090020479f5
> FooId: 2514bc34-f5d6-4833-b14b-5090020479f5
> FooId: 3e1a5a61-c90c-4e67-bfc6-3c3151933c34
> FooId: 3e1a5a61-c90c-4e67-bfc6-3c3151933c34

Its expected that there should be 1 scoped lifetime in one http request, where the "Middleware" and "Function" scopes are both shared.

So this is a bug too, correct?

Im using:
> Core Tools Version:       4.0.4829 Commit hash: N/A  (64-bit)
> Function Runtime Version: 4.11.2.19273

It happens both with .NET 7 rc 2 and .NET 6 with the latest versions of the nuget packages and also the preview packages.

### Workaround
My workout right now, is to instantiate all services in my middleware through the FunctionContext, and then it will work
```
public class MyMiddleware : IFunctionsWorkerMiddleware
{
    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
    {
        var fooService = context.InstanceServices.GetRequiredService<FooService>();
        fooService.LogFooId();
        await next(context);
    }
}
public class MyMiddleware2 : IFunctionsWorkerMiddleware
{
    public async Task Invoke(FunctionContext context, FunctionExecutionDelegate next)
    {
        var fooService = context.InstanceServices.GetRequiredService<FooService>();
        fooService.LogFooId();
        await next(context);
    }
}
```

Then when i call the api, it will log out:
```
FooId: 7b36df46-8ec2-41c8-97f7-57a642ac743f
FooId: 7b36df46-8ec2-41c8-97f7-57a642ac743f
FooId: 7b36df46-8ec2-41c8-97f7-57a642ac743f
FooId: 7b36df46-8ec2-41c8-97f7-57a642ac743f
```

**Edit:**
I have opened this issue in the azure-functions-dotnet-workerproject
https://github.com/Azure/azure-functions-dotnet-worker/issues/1133

**Comment by kshyju** [comment]
@Assassinbeast Your issues seems to be in the isolated model. Could you please open a new issue here: https://github.com/Azure/azure-functions-dotnet-worker/issues

**Comment by fabiocav** [comment]
Closing this issue as based on the recent reports, the default behavior in 4.0 (or using the `EnableEnhancedScopes`) seem to address the problem. 

If you continue to experience similar problems, please open a separate issue with the details so we can take a closer look.

Thank you!
